File: 백준/Silver/1629. 곱셈/곱셈.cc
#include <bits/stdc++.h>
using namespace std;

typedef long long ll;

ll a, b, c;

ll solve(ll a, ll b) {
	if (b == 1) return a % c;

	// 2^4 = 2^2 * 2^2
	ll ret = solve(a, b / 2);
	ret = (ret * ret) % c;
	
	// 2^5 = 2^2 * 2^2 * 2
	if (b % 2 == 1) ret = (ret * a) % c;
	return ret;
}

int main() {
	cin >> a >> b >> c;
	cout << solve(a, b) << "\n";

	return 0;
}

File: 백준/Silver/1629. 곱셈/곱셈.java
import java.io.*;
import java.util.StringTokenizer;

public class Main {

    private static long solve(int a, int b, int c) {
        if (b == 1) return a % c;

        long ret = solve(a, b / 2, c);
        ret = (ret * ret) % c;
        if (b % 2 == 1) ret = (ret * a) % c;

        return ret;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        StringTokenizer st = new StringTokenizer(br.readLine());
        int a = Integer.parseInt(st.nextToken());
        int b = Integer.parseInt(st.nextToken());
        int c = Integer.parseInt(st.nextToken());

        bw.write(solve(a, b, c) + "\n");
        bw.flush();
        br.close();
        bw.close();
    }
}
